[stable/kubernetes-dashboard] - Add support for gce-ingress integration

mbelang opened the pull request:
#### What this PR does / why we need it:
Fix gce-ingress integration for kubernetes-dashboard

- Replace `.Values.ingress.path` with `.Values.ingress.paths` to support gce-ingress that requires `/` and `/*` to work
- Add documentation for gce-ingress (service annotation + ingress paths)
- The service port must have a name to work with gce-ingress

#### Special notes for your reviewer:
This is not backward compatible if you override the default `.Values.ingress.path`.

Addresses https://github.com/kubernetes/contrib/issues/885

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

Commit:
Fix gce-ingress integration for kubernetes-dashboard

To make the gce-ingress working:
- The service needs to have a named port along with annotation for
  the protocol mapping the port name with HTTPS in that case
- The ingress requires both / and /* to be forwarded to the backend
  to be able to serve all the dashboard paths and sub paths

Signed-off-by: Maxime Belanger <mbelanger@dialogue.co>

maver1ck commented on stable/kubernetes-dashboard/templates/svc.yaml:
```diff
@@ -11,15 +11,24 @@ metadata:
 {{- end }}
+{{- if .Values.ingress.enabled }}
+  annotations:
+    service.alpha.kubernetes.io/app-protocols: '{"https":"HTTPS"}'
```
Do we need this ?
We can configure this in annotations

mbelang replied:
It is required to make it work with GCE ingress yes. But I'm not sure I understand what you mean. It is an annotation and configured as such but did you mean via values instead of in the template? if so It would have to be clear that we need that to make it work with GCE ingress. And it is actually to match that: https://github.com/helm/charts/pull/9179/files#diff-87590f01f77a035df78b589038b308d7R31

I had a hard time figuring that out.

maver1ck replied:
I'm not sure how this configuration will work on other service providers

mbelang replied:
Yeah I'm not in a position to test that honestly. If we would make that optional, how is the best approach doing this given it took me a good amount of time to make it work and given I want to save time to other people :). Also google is deprecating the dashboard addon part of GKE so anyone who wants it working on GKE will have that problem.

mbelang replied:
Ok I update the PR @maver1ck. I can do what I want with values except one thing which is the actual fix.

desaintmartin replied:
I would prefer to add a comment in values.yaml / hint in README about `.Values.service.annotations` parameter and how to configure it to have it work with gce, in order to stay generic.

Edit: I did not realize it but this is what you already do, so are those lines still needed?

mbelang committed:
Simplify the change and add comments instead

Signed-off-by: Maxime Belanger <maxime.b.belanger@gmail.com>

mbelang commented on stable/kubernetes-dashboard/values.yaml:
```diff
@@ -80,6 +82,7 @@ ingress:
   ## Kubernetes Dashboard Ingress path
   ##
+  ## For GCE ingress the path should be /*
   path: /
```
Honestly, I do not understand how it is working right now because when I only set `/` as the path, only the requests on `/` route are routed to the dashboard, the rest is routed to default backend. I think `/*` is required but I'm not in position to test that on nginx ingress.

mbelang replied:
Oh I found the answer, nginx ingress and gce ingress doesn't behave equally with paths :(

mbelang commented on stable/kubernetes-dashboard/templates/svc.yaml:
```diff
@@ -20,6 +20,7 @@ spec:
   ports:
   - port: {{ .Values.service.externalPort }}
     targetPort: 8443
+    name: "https"
```
The only real fix required is this one. The rest can be configured via the values.

desaintmartin replied:
Please note that previous PR set an "insecure" access using plain HTTP. Naming it https makes no sense.

mbelang replied:
clearly, missed that.

mbelang committed:
Major changes

- Now supports `paths` in place of `path`
- Add documentation for GCE ingress support instead of adding some logic

Signed-off-by: Maxime Belanger <maxime.b.belanger@gmail.com>

desaintmartin commented on stable/kubernetes-dashboard/README.md:
```diff
@@ -58,7 +58,9 @@ The following table lists the configurable parameters of the kubernetes-dashboar
 | `ingress.enabled`                   | Enable ingress controller resource                                                                                          | `false`                                                                    |
+| `ingress.gce`                       | Enables gce ingress controller specific configurations                                                                      | `false`                                                                    |
 | `ingress.path`                      | Path to match against incoming requests. Must begin with a '/'                                                              | `/`                                                                        |
+| `ingress.paths`                     | Paths to match against incoming requests. This is required to work with gce ingress `/` and `/*` required. Only when `ingress.gce` is true.                 | `[]`                                                                        |
```
Could we simply use this one (defaulting to `["/"]`) and remove the other one? It would make a breaking change but would be more understandable. cc @kfox1111 for opinion.

mbelang replied:
I was tempted to propose a breaking change but I decided to do it that way. If you are welling to do a breaking change I do not mind.

desaintmartin replied:
If @kfox1111 agree I propose:
- `.Values.ingress.paths`, replacing `.Values.ingress.path`. 
- `.Values.ingress.path` is deprecated but still works (`if and .Values.ingress.path (not .Values.ingress.paths)` for example)

It may be worth looking at other charts to see how they handle this.

mbelang replied:
If `.Values.ingress.paths` is defaulted to `[/]` then that condition will never work if someone defines `.Values.ingress.path` which is already defaulted to `/`.

Since this is a major revision, I do not think it is a huge change for somebody to switch. If using default path then the update to `1.0.0` is free. If not, it is only a matter of replacing `path: /` by `paths: [/]`

mbelang committed:
Port naming according to enableInsecureLogin value

Signed-off-by: Maxime Belanger <maxime.b.belanger@gmail.com>

mbelang committed:
Add a migration documentation and fix some .md linting issues

Signed-off-by: Maxime Belanger <maxime.b.belanger@gmail.com>
